### DISCUSSION TITLE
Filter and where function not working

### DISCUSSION THREAD

**Discussion by Anjali-Poornima666** [sqlalchemy/sqlalchemy#8529] [Usage Questions]
I'm using sqlalchemy where functionality to execute a sql query.
my query is : 
select(table_1['part_id'], table_1.columns['location_id'], table_1.columns['value']).where(table_1.columns['part_id'] == "asdfgh")

When I print this clause I get:
SELECT [MS_ENGINE].[RESULTS].part_id, [MS_ENGINE].[RESULTS].location_id, [MS_ENGINE].[RESULTS].value
FROM [MS_ENGINE].[RESULTS]
WHERE [MS_ENGINE].[RESULTS].part_id = ?

My question is, why is there a '?' instead of a parameter like part_id_1

I'm struggling with this issue since a week and I really hope someone helps me!

**Comment by zzzeek** [comment]
When you print a query that is associated with a `MetaData()` object that has `bind=engine` as part of it, the stringification of the query will use the dialect of that engine, which in this case we would assume is pyodbc, which uses question marks for bound parameters.

the solution is to not use `MetaData(bind=engine)`, take the "bind" out, that parameter is deprecated.

Otherwise, you can force "default" stringification like this:

```py
from sqlalchemy.engine.default import DefaultDialect

print(my_statement.compile(dialect=DefaultDialect()))
```

some background at https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#how-do-i-render-sql-expressions-as-strings-possibly-with-bound-parameters-inlined but likely mentions of "bound metadata" have been removed as this is a deprecated case.

**Reply by Anjali-Poornima666** [reply]
Thank you for your solution of compiling with DefaultDialect @zzzeek . But when I execute the complied query, I get the following error:
'The SQL contains 0 parameter markers, but 1 parameters were supplied', 'HY000')

```
part_id = "asdfgh"
s = select(table_1.columns['part_id'], table_1.columns['location_id'], table_1.columns['value']).where(table_1.columns['part_id'] == part_id)

query = s.compile(dialect=DefaultDialect())
query_str = str(query)
query_params = query.params

logging.info(query_str)
logging.info(query_params)
```

The output I get is :
```
SELECT "MS_ENGINE"."RESULTS".part_id, "MS_ENGINE"."RESULTS".location_id, "MS_ENGINE"."RESULTS".value 
FROM "MS_ENGINE"."RESULTS"
WHERE "MS_ENGINE"."RESULTS".part_id = :part_id_1
{'part_id_1': 'asdfgh'}
```
The error I get :

Exception: ProgrammingError: (pyodbc.ProgrammingError) ('The SQL contains 0 parameter markers, but 1 parameters were supplied', 'HY000')
[SQL: SELECT "MS_ENGINE"."RESULTS".part_id, "MS_ENGINE"."RESULTS".location_id, "MS_ENGINE"."RESULTS".value
FROM "MS_ENGINE"."RESULTS"
WHERE "MS_ENGINE"."RESULTS".part_id = :part_id_1]
[parameters: {'part_id_1': 'asdfgh'}]

**Reply by CaselIT** [reply]
You cannot run the default dialect query on pyodbc.

I'm not sure what you are trying to do here. The usual way of running statement with sqlalchemy is not converting them to string and executing the string, but to pass the statement to execute. So `connection.execute(s)`

**Reply by Anjali-Poornima666** [reply]
But when I directly execute the statement, I'm getting an empty result. But when I use the same statement in sql directly, I get the output. What do you think is the problem then? The column I'm filtering is a NVARCHAR. But when I filter on a column with type decimal, I get some result.

**Reply by CaselIT** [reply]
what do you mean with
>  But when I use the same statement in sql directly, I get the output

is this running the sql string via sqlalchemy or using a different tool, like dbeaver, to run the query?

Regarding mssql I'm not too familiar, but I think it should handle the conversion for an `==` condition. You may specify the type of argument by setting a type to it with `sa.literal(param, type_=sa.NVARCHAR)` and optionally using cast if that's still needed

**Reply by Anjali-Poornima666** [reply]
The query I wanted to get the output is :

s = select(table_1.columns['part_id'], table_1.columns['location_id'], table_1.columns['value']).where(table_1.columns['part_id'] == part_id)

1. When I run this using SSMS (run the sql select query), I get the required output. 
2. But when I use sqlalchemy to run this query, I don't get any output.

What do you I have done the mistake?

```
engine = create_engine(connection_string)
connection = engine.connect()
metadata = MetaData(engine)
metadata.reflect()
table_1 = Table('RESULTS', metadata, schema="MS_ENGINE", autoload=True)
Session = sessionmaker(bind=engine)
session = Session()
part_id = "asdfgh"
res = session.query(table_1 .columns['part_id'], table_1 .columns['location_id'], table_1.columns['value']).filter(table_1.c.part_id == part_id).limit(5)
```

I have also tried with : 
```
s = select(table_1.columns['part_id'], table_1.columns['location_id'], table_1.columns['value']).where(table_1.columns['part_id'] == part_id)
ResultProxy = connection.execute(s)
ResultSet = ResultProxy.fetchall()
logging.info(ResultSet[:10])
```
But, nothing works. Do you have any suggestions?

**Reply by CaselIT** [reply]
that looks strange.
you can enable echo=True in create_engine to make it log what queries are executed.

are you sure that the db is the same?
what's the type of  the `part_id` column?

**Reply by Anjali-Poornima666** [reply]
Yes, the db is the same. I'm confident that the db and tables are right because when I filter using value column, I get the output. But when I filter using part_id column, I don't get any output. 

The type of `value` column is decimal and type of `part_id `column is NVARCHAR.
Please try to help me with some solution!

**Reply by CaselIT** [reply]
this seems like mssql that does things randomly and really require a cast.
try changing the query to use `table_1.columns['part_id'] == sa.cast(part_id, NVARCHAR)` or `table_1.columns['part_id'] == sa.literal(part_id, NVARCHAR)`

@zzzeek any idea why mssql/pyodc behaves so poorly here?

**Reply by Anjali-Poornima666** [reply]
Sorry, but none of them worked! Any other suggestion or trick.

**Reply by CaselIT** [reply]
This really looks like a different db/table is being queried, since none of there issues has ever been reported

What does the log prints when passing `echo=True` to create engine when executing the query with the above cast?

**Reply by Anjali-Poornima666** [reply]
I have passed the argument echo=True to create engine. Following are the logs : 
1. When I run the filter with `value` column:
```

2022-09-21 14:56:05,235 INFO sqlalchemy.engine.Engine BEGIN (implicit)
[2022-09-21T12:56:05.238Z] BEGIN (implicit)
[2022-09-21T12:56:05.238Z] 2022-09-21 14:56:05,237 INFO sqlalchemy.engine.Engine SELECT TOP 5 [MS_ENGINE].[RESULTS].part_id AS [MS_ENGINE_RESULTS_part_id], [MS_ENGINE].[RESULTS].location_id AS [MS_ENGINE_RESULTS_location_id], [MS_ENGINE].[RESULTS].value AS [MS_ENGINE_RESULTS_value]
[2022-09-21T12:56:05.240Z] SELECT TOP 5 [MS_ENGINE].[RESULTS].part_id AS [MS_ENGINE_RESULTS_part_id], [MS_ENGINE].[RESULTS].location_id AS [MS_ENGINE_RESULTS_location_id], [MS_ENGINE].[RESULTS].value AS [MS_ENGINE_RESULTS_value] 
FROM [MS_ENGINE].[RESULTS]
WHERE [MS_ENGINE].[RESULTS].value = ?
[2022-09-21T12:56:05.243Z] FROM [MS_ENGINE].[RESULTS] 
[2022-09-21T12:56:05.247Z] [generated in 0.00069s] ('0.2480',)
[2022-09-21T12:56:05.248Z] WHERE [MS_ENGINE].[RESULTS].value = ?
[2022-09-21T12:56:05.251Z] 2022-09-21 14:56:05,238 INFO sqlalchemy.engine.Engine [generated in 0.00069s] ('0.2480',)
```

2. When I run the filter with `part_id` column:
```
2022-09-21 14:58:14,775 INFO sqlalchemy.engine.Engine BEGIN (implicit)
[2022-09-21T12:58:14.778Z] BEGIN (implicit)
[2022-09-21T12:58:14.778Z] 2022-09-21 14:58:14,778 INFO sqlalchemy.engine.Engine SELECT TOP 5 [MS_ENGINE].[RESULTS].part_id AS [MS_ENGINE_RESULTS_part_id], [MS_ENGINE].[RESULTS].location_id AS [MS_ENGINE_RESULTS_location_id], [MS_ENGINE].[RESULTS].value AS [MS_ENGINE_RESULTS_value]
[2022-09-21T12:58:14.779Z] SELECT TOP 5 [MS_ENGINE].[RESULTS].part_id AS [MS_ENGINE_RESULTS_part_id], [MS_ENGINE].[RESULTS].location_id AS [MS_ENGINE_RESULTS_location_id],[MS_ENGINE].[RESULTS].value AS [MS_ENGINE_RESULTS_value]
FROM [MS_ENGINE].[RESULTS]
WHERE [MS_ENGINE].[RESULTS].part_id = ?
[2022-09-21T12:58:14.781Z] FROM [MS_ENGINE].[RESULTS] 
[2022-09-21T12:58:14.783Z] [generated in 0.00088s] ('AB 12345',)
[2022-09-21T12:58:14.784Z] WHERE [MS_ENGINE].[RESULTS].part_id = ?
[2022-09-21T12:58:14.785Z] 2022-09-21 14:58:14,778 INFO sqlalchemy.engine.Engine [generated in 0.00088s] ('AB 12345',)
```
Seeing the above logs, I don't find any difference except the where clause. 
Some additional information, I'm using this to set an API using Azure functions and the database is a Azure sql database.

**Reply by zzzeek** [reply]
im not totally following what the problem is here, but one thing that looks funny is the comparison of "value" to a string.  is value a numeric column?  (Seems like it's decimal) so OK if you are using the Numeric type, you'd pass a Python Decimal object, not a string.  If it's a Float type, pass a Python floating point value, not a string.

**Reply by zzzeek** [reply]
here's how we can know what the issue is.  go to your SQL Server command line, and write the query that you want, run it and make sure it gets the results you want.  show us that query.

**Reply by Anjali-Poornima666** [reply]
The query I run in SQL server command line is : 

```
SELECT TOP 5
      [part_id]
      ,[location_id]
      ,[value]
     
  FROM [MS_ENGINE].[RESULTS] 
  WHERE [part_id]='AB 12345' 
```

**Reply by zzzeek** [reply]
OK that matches, let's go to the driver and see if it works from there:

```py
with engine.connect() as conn:
    pyodbc_connection = conn.connection
    cursor = pyodbc.cursor()
    cursor.execute("""
 SELECT TOP 5 [MS_ENGINE].[RESULTS].part_id AS [MS_ENGINE_RESULTS_part_id], [MS_ENGINE].[RESULTS].location_id AS [MS_ENGINE_RESULTS_location_id],[MS_ENGINE].[RESULTS].value AS [MS_ENGINE_RESULTS_value]
FROM [MS_ENGINE].[RESULTS]
WHERE [MS_ENGINE].[RESULTS].part_id = ?
""", ('AB 12345',)
)
    print(cursor.fetchall())
    cursor.close()
```

that's the pyodbc cursor. you want to get that part working first since that's where we go.   if there is some weird collation problem etc

**Reply by Anjali-Poornima666** [reply]
Hello @zzzeek and @CaselIT, Thank you so much for your time and inputs. Yesterday, after you mentioned about the database. I went back to my cloud architect to confirm if the database I'm using in SSMS and the connection string in python is the same or not. He confirmed that they are not the same and I have to change accordingly. And now the query works as expected. 

Once again. sorry for the trouble caused and Thanks a lot for your inputs. :)
